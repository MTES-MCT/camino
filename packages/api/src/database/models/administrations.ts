import { Model } from 'objection'

import { IAdministration } from '../../types'

import AdministrationsTitresTypesEtapesTypes from './administrations-titres-types-etapes-types'
import AdministrationsTitresTypesTitresStatuts from './administrations-titres-types-titres-statuts'
import Utilisateurs from './utilisateurs'
import Titres from './titres'
import ActivitesTypes from './activites-types'

interface Administrations extends IAdministration {}

class Administrations extends Model {
  public static tableName = 'administrations'

  public static jsonSchema = {
    type: 'object',
    required: ['id'],

    properties: {
      id: { type: 'string', maxLength: 64 }
    }
  }

  static relationMappings = () => ({
    titresTypesTitresStatuts: {
      relation: Model.HasManyRelation,
      modelClass: AdministrationsTitresTypesTitresStatuts,
      join: {
        from: 'administrations.id',
        to: 'administrations__titresTypes__titresStatuts.administrationId'
      }
    },

    titresTypesEtapesTypes: {
      relation: Model.HasManyRelation,
      modelClass: AdministrationsTitresTypesEtapesTypes,
      join: {
        from: 'administrations.id',
        to: 'administrations__titresTypes__etapesTypes.administrationId'
      }
    },

    utilisateurs: {
      relation: Model.HasManyRelation,
      modelClass: Utilisateurs,
      join: {
        from: 'administrations.id',
        to: 'utilisateurs.administrationId'
      }
    },

    gestionnaireTitres: {
      relation: Model.ManyToManyRelation,
      modelClass: Titres,
      join: {
        from: 'administrations.id',
        through: {
          from: 'titresAdministrationsGestionnaires.administrationId',
          to: 'titresAdministrationsGestionnaires.titreId'
        },
        to: 'titres.id'
      }
    },

    activitesTypes: {
      relation: Model.ManyToManyRelation,
      modelClass: ActivitesTypes,
      join: {
        from: 'administrations.id',
        through: {
          from: 'administrations__activitesTypes.administrationId',
          to: 'administrations__activitesTypes.activiteTypeId',
          extra: ['modificationInterdit', 'lectureInterdit']
        },
        to: 'activitesTypes.id'
      }
    },

    activitesTypesEmails: {
      relation: Model.ManyToManyRelation,
      modelClass: ActivitesTypes,
      join: {
        from: 'administrations.id',
        through: {
          from: 'administrations__activitesTypes__emails.administrationId',
          to: 'administrations__activitesTypes__emails.activiteTypeId',
          extra: ['email']
        },
        to: 'activitesTypes.id'
      }
    }
  })
}

export default Administrations
