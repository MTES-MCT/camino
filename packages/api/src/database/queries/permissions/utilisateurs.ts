import { QueryBuilder } from 'objection'

import Utilisateurs from '../../models/utilisateurs'
import Entreprises from '../../models/entreprises'
import { entreprisesQueryModify } from './entreprises'
import { isAdministrationEditeur, isAdministrationLecteur, isBureauDEtudes, isDefault, isEntreprise, User } from 'camino-common/src/roles'

export const utilisateursQueryModify = (q: QueryBuilder<Utilisateurs, Utilisateurs | Utilisateurs[]>, user: User) => {
  q.select('utilisateurs.*')

  q.whereNotNull('utilisateurs.keycloakId')

  if (isAdministrationEditeur(user) || isAdministrationLecteur(user)) {
    // un utilisateur 'editeur' ou 'lecteur'
    // ne voit que les utilisateurs de son administration
    q.where('utilisateurs.administrationId', user.administrationId)
  } else if ((isEntreprise(user) || isBureauDEtudes(user)) && user.entreprises.length) {
    // un utilisateur entreprise
    // ne voit que les utilisateurs de son entreprise
    const entreprisesIds = user.entreprises.map(e => e.id)

    q.whereExists((Utilisateurs.relatedQuery('entreprises') as QueryBuilder<Entreprises, Entreprises | Entreprises[]>).whereIn('entreprises.id', entreprisesIds))
  } else if (user && isDefault(user)) {
    // un utilisateur "defaut" ne voit que son propre profil
    q.where('id', user.id)
  } else if (!user) {
    // un utilisateur non-authentifiÃ© ne voit aucun utilisateur
    q.where(false)
  }

  q.modifyGraph('entreprises', u => entreprisesQueryModify(u as QueryBuilder<Entreprises, Entreprises | Entreprises[]>, user))

  return q
}
