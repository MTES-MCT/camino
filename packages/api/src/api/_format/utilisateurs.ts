import { IUtilisateur } from '../../types'
import { administrationFormat } from './administrations'

export const utilisateurFormat = (utilisateur: IUtilisateur) => {
  // TODO: devrait formater les entreprises  de l'utilisateur

  if (utilisateur.administrations) {
    utilisateur.administrations =
      utilisateur.administrations.map(administrationFormat)
  }

  return utilisateur
}
