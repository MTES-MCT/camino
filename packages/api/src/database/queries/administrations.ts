import {
  IAdministrationActiviteType,
  IAdministrationActiviteTypeEmail,
  IAdministrationTitreTypeEtapeType,
  IAdministrationTitreTypeTitreStatut,
  IFields,
  IUtilisateur
} from '../../types.js'

import graphBuild from './graph/build.js'
import { fieldsFormat } from './graph/fields-format.js'
import options from './_options.js'

import Administrations from '../models/administrations.js'
import { administrationsQueryModify } from './permissions/administrations.js'
import AdministrationsTitresTypesTitresStatuts from '../models/administrations-titres-types-titres-statuts.js'
import AdministrationsTitresTypesEtapesTypes from '../models/administrations-titres-types-etapes-types.js'
import AdministrationsActivitesTypes from '../models/administrations-activites-types.js'
import AdministrationsActivitesTypesEmails from '../models/administrations-activites-types-emails.js'

const administrationsQueryBuild = (
  { fields }: { fields?: IFields },
  user: IUtilisateur | null | undefined
) => {
  const graph = fields
    ? graphBuild(fields, 'administrations', fieldsFormat)
    : options.administrations.graph

  const q = Administrations.query().withGraphFetched(graph)

  administrationsQueryModify(q, user)

  return q
}

const administrationGet = async (
  id: string,
  { fields }: { fields?: IFields },
  user: IUtilisateur | null | undefined
) => {
  const q = administrationsQueryBuild({ fields }, user)

  return q.findById(id)
}

const administrationsGet = async (
  { fields }: { fields?: IFields },
  user: IUtilisateur | null | undefined
) => {
  return administrationsQueryBuild({ fields }, user)
}

const administrationTitreTypeTitreStatutUpsert = async (
  administrationTitreTypeTitreStatut: IAdministrationTitreTypeTitreStatut
) =>
  AdministrationsTitresTypesTitresStatuts.query().upsertGraph(
    administrationTitreTypeTitreStatut,
    { insertMissing: true }
  )

const administrationTitreTypeTitreStatutDelete = async (
  administrationId: string,
  titreTypeId: string,
  statutTypeId: string
) =>
  AdministrationsTitresTypesTitresStatuts.query().deleteById([
    administrationId,
    titreTypeId,
    statutTypeId
  ])

const administrationTitreTypeEtapeTypeUpsert = async (
  administrationTitreTypeEtapeType: IAdministrationTitreTypeEtapeType
) =>
  AdministrationsTitresTypesEtapesTypes.query().upsertGraph(
    administrationTitreTypeEtapeType,
    { insertMissing: true }
  )

const administrationTitreTypeEtapeTypeDelete = async (
  administrationId: string,
  titreTypeId: string,
  etapeTypeId: string
) =>
  AdministrationsTitresTypesEtapesTypes.query().deleteById([
    administrationId,
    titreTypeId,
    etapeTypeId
  ])

const administrationActiviteTypeUpsert = async (
  administrationActiviteType: IAdministrationActiviteType
) =>
  AdministrationsActivitesTypes.query().upsertGraph(
    administrationActiviteType,
    {
      insertMissing: true
    }
  )

const administrationActiviteTypeDelete = async (
  administrationId: string,
  ActiviteTypeId: string
) =>
  AdministrationsActivitesTypes.query().deleteById([
    administrationId,
    ActiviteTypeId
  ])

const administrationActiviteTypeEmailCreate = async (
  administrationActiviteTypeEmail: IAdministrationActiviteTypeEmail
) =>
  AdministrationsActivitesTypesEmails.query().insertGraph(
    administrationActiviteTypeEmail
  )

const administrationActiviteTypeEmailDelete = async (
  administrationActiviteTypeEmail: IAdministrationActiviteTypeEmail
) => {
  const { activiteTypeId, administrationId, email } =
    administrationActiviteTypeEmail

  return AdministrationsActivitesTypesEmails.query().delete().where({
    activite_type_id: activiteTypeId,
    administration_id: administrationId,
    email
  })
}

export {
  administrationGet,
  administrationsGet,
  administrationTitreTypeTitreStatutUpsert,
  administrationTitreTypeTitreStatutDelete,
  administrationTitreTypeEtapeTypeUpsert,
  administrationTitreTypeEtapeTypeDelete,
  administrationActiviteTypeUpsert,
  administrationActiviteTypeDelete,
  administrationActiviteTypeEmailCreate,
  administrationActiviteTypeEmailDelete
}
