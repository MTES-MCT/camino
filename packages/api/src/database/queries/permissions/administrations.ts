import { raw, QueryBuilder } from 'objection'

import { IUtilisateur } from '../../../types'

import { knex } from '../../../knex'

import AdministrationsModel from '../../models/administrations'
import Utilisateurs from '../../models/utilisateurs'
import Titres from '../../models/titres'

import { titresQueryModify } from './titres'
import { utilisateursQueryModify } from './utilisateurs'
import {
  isAdministrationAdmin,
  isAdministrationEditeur,
  isSuper
} from 'camino-common/src/roles'
import {
  AdministrationId,
  Administrations,
  sortedAdministrations
} from 'camino-common/src/static/administrations'
import { Departements } from 'camino-common/src/static/departement'
import { getTitreTypeIdsByAdministration } from 'camino-common/src/static/administrationsTitresTypes'
import { getKeys } from 'camino-common/src/typescript-tools'

const administrationsQueryModify = (
  q: QueryBuilder<
    AdministrationsModel,
    AdministrationsModel | AdministrationsModel[]
  >,
  user: IUtilisateur | null | undefined
): QueryBuilder<
  AdministrationsModel,
  AdministrationsModel | AdministrationsModel[]
> => {
  q.select('administrations.*')

  if (isSuper(user)) {
    q.select(raw('true').as('modification'))
  }

  if (
    isSuper(user) ||
    ((isAdministrationAdmin(user) || isAdministrationEditeur(user)) &&
      Administrations[user.administrationId].typeId === 'min')
  ) {
    // Utilisateur super ou membre de ministère (admin ou éditeur) : tous les droits
    q.select(raw('true').as('emailsModification'))
  } else if (isAdministrationAdmin(user) || isAdministrationEditeur(user)) {
    // Membre d'une DREAL/DEAL vis-à-vis de la DREAL elle-même,
    // ou d'un DREAL/DEAL vis-à-vis d'une administration qui dépend d'elles
    // Admin ou éditeur : modifications
    // Admin, éditeur ou lecteur : lecture

    const administration = Administrations[user.administrationId]
    if (administration.regionId) {
      const departementIds = Object.values(Departements)
        .filter(({ regionId }) => regionId === administration.regionId)
        .map(({ id }) => id)
      // On récupère toutes les administrations départementales qui sont de la même région que l’administration régionale de l’utilisateur
      const administrationIds = sortedAdministrations
        .filter(({ departementId }) => departementIds.includes(departementId))
        .map(({ id }) => id)

      if (administration.typeId === 'dre' || administration.typeId === 'dea') {
        administrationIds.push(user.administrationId)
      }
      if (administrationIds.length) {
        q.select(
          raw(
            `administrations.id IN (${administrationIds
              .map(() => '?')
              .join(',')})`,
            administrationIds
          ).as('emailsModification')
        )
      }
    }
  }

  q.modifyGraph('gestionnaireTitres', a =>
    titresQueryModify(a as QueryBuilder<Titres, Titres | Titres[]>, user)
      // on group by administrationId au cas où il y a une aggrégation
      // dans la requête de titre (ex : calc activités)
      .groupBy(
        'titres.id',
        'titresAdministrationsGestionnaires.administrationId'
      )
  )

  q.modifyGraph('localeTitres', a =>
    titresQueryModify(a as QueryBuilder<Titres, Titres | Titres[]>, user)
      // on group by administrationId au cas où il y a une aggrégation
      // dans la requête de titre (ex : calc activités)
      .groupBy('titres.id', 'titresAdministrationsLocales.administrationId')
  )

  q.modifyGraph('utilisateurs', b => {
    utilisateursQueryModify(
      b as QueryBuilder<Utilisateurs, Utilisateurs | Utilisateurs[]>,
      user
    )
  })

  return q
}

const administrationsLocalesModify = (
  q: QueryBuilder<
    AdministrationsModel,
    AdministrationsModel | AdministrationsModel[]
  >,
  administrationId: AdministrationId,
  titreAlias: string
) => {
  q.leftJoin('titresAdministrationsLocales as t_al', b => {
    b.on(
      knex.raw('?? ->> ? = ??', [
        `${titreAlias}.propsTitreEtapesIds`,
        'administrations',
        't_al.titreEtapeId'
      ])
    )
    b.on(knex.raw('?? = ?', ['t_al.administrationId', administrationId]))
  })
}

const administrationsActivitesModify = (
  q: QueryBuilder<
    AdministrationsModel,
    AdministrationsModel | AdministrationsModel[]
  >,

  { lecture, modification }: { lecture?: boolean; modification?: boolean }
) => {
  q.leftJoin('administrations__activitesTypes as a_at', b => {
    b.on(knex.raw('?? = ??', ['a_at.administrationId', 'administrations.id']))
    b.andOn(
      knex.raw('?? = ??', ['a_at.activiteTypeId', 'titresActivites.typeId'])
    )
    b.andOn(c => {
      if (lecture) {
        c.orOn(knex.raw('?? is true', ['a_at.lectureInterdit']))
      }
      if (modification) {
        c.orOn(knex.raw('?? is true', ['a_at.modificationInterdit']))
      }
    })
  })
  q.whereNull('a_at.administrationId')
}

const administrationsTitresQuery = (
  administrationId: AdministrationId,
  titreAlias: string,
  {
    isGestionnaire,
    isAssociee,
    isLocale
  }: { isGestionnaire?: boolean; isAssociee?: boolean; isLocale?: boolean } = {}
) => {
  const q = AdministrationsModel.query().where(
    'administrations.id',
    administrationId
  )

  if (isLocale) {
    q.modify(administrationsLocalesModify, administrationId, titreAlias)
  }

  q.where(c => {
    if (isGestionnaire || isAssociee) {
      const titresTypesByAdministration =
        getTitreTypeIdsByAdministration(administrationId)
      if (titresTypesByAdministration) {
        const titreTypeIds = getKeys(titresTypesByAdministration).filter(
          titreTypeId => {
            if (
              isGestionnaire &&
              titresTypesByAdministration[titreTypeId]?.gestionnaire
            ) {
              return true
            }
            if (
              isAssociee &&
              titresTypesByAdministration[titreTypeId]?.associee
            ) {
              return true
            }

            return false
          }
        )

        if (titreTypeIds.length) {
          c.orWhereRaw(`?? in (${titreTypeIds.map(t => `'${t}'`).join(',')})`, [
            `${titreAlias}.typeId`
          ])
        } else {
          c.orWhereRaw('false')
        }
      } else {
        c.orWhereRaw('false')
      }
    }

    if (isLocale) {
      c.orWhereNotNull('t_al.administrationId')
    }
  })

  return q
}

const administrationsTitresTypesTitresStatutsModify = (
  q: QueryBuilder<
    AdministrationsModel,
    AdministrationsModel | AdministrationsModel[]
  >,
  type: 'titres' | 'demarches' | 'etapes',
  titreAlias: string,
  conditionsAdd?: (
    b: QueryBuilder<
      AdministrationsModel,
      AdministrationsModel | AdministrationsModel[]
    >
  ) => void
) => {
  q.leftJoin('administrations__titresTypes__titresStatuts as a_tt_ts', b => {
    b.on(
      knex.raw('?? = ??', ['a_tt_ts.administrationId', 'administrations.id'])
    )
    b.andOn(
      knex.raw('?? = ??', ['a_tt_ts.titreTypeId', `${titreAlias}.typeId`])
    )
    b.andOn(
      knex.raw('?? = ??', [
        'a_tt_ts.titreStatutId',
        `${titreAlias}.titreStatutId`
      ])
    )
    b.andOn(knex.raw('?? is true', [`a_tt_ts.${type}ModificationInterdit`]))
  })

  q.where(b => {
    b.orWhereNull('a_tt_ts.administrationId')
    if (conditionsAdd) {
      conditionsAdd(b)
    }
  })
}

// l'utilisateur est dans au moins une administration
// qui n'a pas de restriction 'creationInterdit' sur ce type d'étape / type de titre
const administrationsTitresTypesEtapesTypesModify = (
  q: QueryBuilder<
    AdministrationsModel,
    AdministrationsModel | AdministrationsModel[]
  >,
  type: 'lecture' | 'modification' | 'creation',
  titreTypeIdColumn: string,
  etapeTypeIdColumn: string
) => {
  q.leftJoin('administrations__titresTypes__etapesTypes as a_tt_et', b => {
    b.on(
      knex.raw('?? = ??', ['a_tt_et.administrationId', 'administrations.id'])
    )
    b.andOn(knex.raw('?? = ??', ['a_tt_et.titreTypeId', titreTypeIdColumn]))
    b.andOn(knex.raw('?? = ??', ['a_tt_et.etapeTypeId', etapeTypeIdColumn]))
    b.andOn(knex.raw('?? is true', [`a_tt_et.${type}Interdit`]))
  }).whereNull('a_tt_et.administrationId')
}

export {
  administrationsQueryModify,
  administrationsLocalesModify,
  administrationsTitresTypesTitresStatutsModify,
  administrationsTitresTypesEtapesTypesModify,
  administrationsTitresQuery,
  administrationsActivitesModify
}
