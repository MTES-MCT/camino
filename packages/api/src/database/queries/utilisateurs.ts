import { RawBuilder, QueryBuilder } from 'objection'

import { IUtilisateursColonneId, IColonne, IUtilisateurTitre, IUtilisateur, formatUser } from '../../types'

import options, { FieldsUtilisateur, FieldsUtilisateurTitre } from './_options'
import graphBuild from './graph/build'
import { fieldsFormat } from './graph/fields-format'
import { stringSplit } from './_utils'

import Utilisateurs from '../models/utilisateurs'
import { utilisateursQueryModify } from './permissions/utilisateurs'
import UtilisateursTitres from '../models/utilisateurs--titres'
import { Role, User } from 'camino-common/src/roles'
import { DeepReadonly, isNotNullNorUndefined, isNotNullNorUndefinedNorEmpty, isNullOrUndefined } from 'camino-common/src/typescript-tools'

const userGet = async (userId?: string): Promise<User> => {
  if (isNullOrUndefined(userId)) return null

  const user = await Utilisateurs.query().withGraphFetched('[entreprises]').findById(userId)
  if (user) {
    const q = utilisateursQueryBuild(
      {
        fields: {
          entreprises: { id: {} },
        },
      },
      formatUser(user)
    )

    const userBdd: IUtilisateur | undefined = await q.findById(userId)
    if (userBdd) {
      return formatUser(userBdd)
    }
  }

  return undefined
}

const utilisateursQueryBuild = ({ fields }: { fields?: FieldsUtilisateur }, user: DeepReadonly<User>) => {
  const graph = fields ? graphBuild(fields, 'utilisateur', fieldsFormat) : options.utilisateurs.graph

  const q = Utilisateurs.query().withGraphFetched(graph)

  utilisateursQueryModify(q, user)

  return q
}

const utilisateursFiltersQueryModify = (
  {
    ids,
    entreprisesIds,
    administrationIds,
    roles,
    noms,
    emails,
  }: {
    ids?: string[]
    entreprisesIds?: string[]
    administrationIds?: string[]
    roles?: Role[]
    noms?: string | null
    emails?: string | null
  },
  q: QueryBuilder<Utilisateurs, Utilisateurs[]>
) => {
  if (ids && ids.length > 0) {
    q.whereIn('id', ids)
  }

  if (roles && roles.length > 0) {
    q.whereIn('role', roles as string[])
  }

  if (administrationIds && administrationIds.length > 0) {
    q.whereIn('administrationId', administrationIds)
  }

  if (entreprisesIds && entreprisesIds.length > 0) {
    q.whereIn('entreprises.id', entreprisesIds).leftJoinRelated('entreprises')
  }

  if (isNotNullNorUndefinedNorEmpty(noms)) {
    const nomsArray = stringSplit(noms)
    const fields = ['nom', 'prenom']

    nomsArray.forEach(s => {
      q.where(b => {
        fields.forEach(f => {
          b.orWhereRaw(`lower(??) like ?`, [`utilisateurs.${f}`, `%${s.toLowerCase()}%`])
        })
      })
    })
  }

  if (isNotNullNorUndefinedNorEmpty(emails)) {
    q.where(b => {
      b.whereRaw(`LOWER(??) LIKE LOWER(?)`, ['utilisateurs.email', `%${emails}%`])
    })
  }

  return q
}

const userByEmailGet = async (email: string | null | undefined): Promise<Utilisateurs | undefined> => {
  if (isNotNullNorUndefinedNorEmpty(email)) {
    const user: IUtilisateur | undefined = await Utilisateurs.query().withGraphFetched('[entreprises]').where('utilisateurs.email', email).first()

    if (user) {
      const q = utilisateursQueryBuild(
        {
          fields: {
            entreprises: { id: {} },
          },
        },
        formatUser(user)
      )

      return q.findById(user.id)
    }
  }

  return undefined
}

export const userByKeycloakIdGet = async (keycloakId: string | null | undefined): Promise<IUtilisateur | undefined> => {
  if (isNotNullNorUndefinedNorEmpty(keycloakId)) {
    const user: IUtilisateur | undefined = await Utilisateurs.query().withGraphFetched('[entreprises]').where('utilisateurs.keycloakId', keycloakId).first()

    if (user) {
      const q = utilisateursQueryBuild(
        {
          fields: {
            entreprises: { id: {} },
          },
        },
        formatUser(user)
      )

      return q.findById(user.id)
    }
  }

  return undefined
}

const utilisateurGet = async (id: string, { fields }: { fields?: FieldsUtilisateur } = {}, user: User) => {
  const q = utilisateursQueryBuild({ fields }, user)

  return q.findById(id)
}

// lien = administration ou entreprise(s) en relation avec l'utilisateur :
// on trie sur la concaténation du nom de l'administration
// avec l'aggrégation ordonnée(STRING_AGG) des noms des entreprises
const utilisateursColonnes: Record<IUtilisateursColonneId, IColonne<string | RawBuilder>> = {
  nom: { id: 'nom' },
  prenom: { id: 'prenom' },
  email: { id: 'email' },
  role: { id: 'role' },
}
const utilisateursGet = async (
  {
    intervalle,
    page,
    colonne,
    ordre,
    ids,
    entreprisesIds,
    administrationIds,
    roles,
    noms,
    emails,
  }: {
    intervalle?: number | null
    page?: number | null
    colonne?: IUtilisateursColonneId | null
    ordre?: 'asc' | 'desc' | null
    ids?: string[]
    entreprisesIds?: string[]
    administrationIds?: string[]
    roles?: Role[]
    noms?: string | null
    emails?: string | null
  },
  { fields }: { fields?: FieldsUtilisateur } = {},
  user: DeepReadonly<User>
) => {
  const q = utilisateursQueryBuild({ fields }, user)

  utilisateursFiltersQueryModify(
    {
      ids,
      entreprisesIds,
      administrationIds,
      roles,
      noms,
      emails,
    },
    q
  )

  if (colonne) {
    q.orderBy(utilisateursColonnes[colonne].id, ordre || 'asc')
  } else {
    q.orderBy('utilisateurs.nom', 'asc')
  }

  if (isNotNullNorUndefined(page) && page > 0 && isNotNullNorUndefined(intervalle) && intervalle > 0) {
    q.offset((page - 1) * intervalle)
  }

  if (isNotNullNorUndefined(intervalle) && intervalle > 0) {
    q.limit(intervalle)
  }

  return q
}

const utilisateursCount = async (
  {
    ids,
    entreprisesIds,
    administrationIds,
    roles,
    noms,
    emails,
  }: {
    ids?: string[]
    entreprisesIds?: string[]
    administrationIds?: string[]
    roles?: Role[]
    noms?: string | null
    emails?: string | null
  },
  { fields }: { fields?: FieldsUtilisateur },
  user: DeepReadonly<User>
) => {
  const q = utilisateursQueryBuild({ fields }, user)

  utilisateursFiltersQueryModify({ ids, entreprisesIds, administrationIds, roles, noms, emails }, q)

  return q.resultSize()
}

const utilisateurCreate = async (utilisateur: IUtilisateur, { fields }: { fields?: FieldsUtilisateur }) =>
  Utilisateurs.query()
    .insertGraph(utilisateur, options.utilisateurs.update)
    .withGraphFetched(fields ? graphBuild(fields, 'utilisateur', fieldsFormat) : options.utilisateurs.graph)
    .first()

const utilisateurUpsert = async (utilisateur: Pick<IUtilisateur, 'id'> & Partial<Omit<IUtilisateur, 'id'>>) => Utilisateurs.query().upsertGraph(utilisateur, options.utilisateurs.update)

const utilisateurTitreCreate = async (utilisateurTitre: IUtilisateurTitre) => UtilisateursTitres.query().insert(utilisateurTitre)

const utilisateurTitreDelete = async (utilisateurId: string, titreId: string) => UtilisateursTitres.query().deleteById([utilisateurId, titreId])

const utilisateursTitresGet = async (titreId: string, { fields }: { fields?: FieldsUtilisateurTitre }) =>
  UtilisateursTitres.query()
    .where('titreId', titreId)
    .withGraphFetched(fields ? graphBuild(fields, 'utilisateursTitres', fieldsFormat) : options.utilisateursTitres.graph)

export { userGet, utilisateurGet, userByEmailGet, utilisateursGet, utilisateursCount, utilisateurCreate, utilisateurUpsert, utilisateurTitreCreate, utilisateurTitreDelete, utilisateursTitresGet }
