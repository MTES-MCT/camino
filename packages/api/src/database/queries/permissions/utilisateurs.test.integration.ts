import { IUtilisateur } from '../../../types'

import { dbManager } from '../../../../tests/db-manager'
import Utilisateurs from '../../models/utilisateurs'
import { utilisateursGet } from '../utilisateurs'
import { Administrations } from 'camino-common/src/administrations'
import options from '../_options'

console.info = jest.fn()
console.error = jest.fn()
beforeAll(async () => {
  await dbManager.populateDb()
  await Utilisateurs.query().insertGraph(mockUser, options.utilisateurs.update)
})

afterAll(async () => {
  await dbManager.closeKnex()
})

const mockAdministration = Administrations['aut-97300-01']

const mockUser: IUtilisateur = {
  id: 'utilisateurId',
  role: 'editeur',
  nom: 'utilisateurNom',
  email: 'utilisateurEmail',
  motDePasse: 'utilisateurMotdepasse',
  administrationId: mockAdministration.id,
  dateCreation: '2022-05-12'
}

describe('utilisateursQueryModify', () => {
  test.each`
    role            | voit
    ${'super'}      | ${true}
    ${'admin'}      | ${true}
    ${'editeur'}    | ${true}
    ${'lecteur'}    | ${true}
    ${'entreprise'} | ${true}
    ${'defaut'}     | ${false}
  `(
    "Vérifie l'écriture de la requête sur un utilisateur",
    async ({ role, voit }) => {
      const user: IUtilisateur = {
        id: 'userId',
        role,
        administrationId: mockAdministration.id,
        dateCreation: '2022-05-12'
      }

      const utilisateurs = await utilisateursGet(
        { noms: mockUser.nom },
        {},
        user
      )
      if (voit) {
        expect(utilisateurs).toHaveLength(1)
        expect(utilisateurs[0]).toMatchSnapshot()
      } else {
        expect(utilisateurs).toHaveLength(0)
      }
    }
  )
})
