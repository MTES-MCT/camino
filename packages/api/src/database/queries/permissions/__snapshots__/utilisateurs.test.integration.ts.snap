// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`utilisateursQueryModify Vérifie l'écriture de la requête sur un utilisateur 1`] = `
Object {
  "administration": Object {
    "activitesTypes": Array [],
    "departementId": null,
    "id": "aut-97300-01",
    "regionId": null,
    "titresTypes": Array [
      Object {
        "archive": null,
        "associee": true,
        "contenuIds": Array [
          Object {
            "elementId": "mecanise",
            "sectionId": "arm",
          },
          Object {
            "elementId": "franchissements",
            "sectionId": "arm",
          },
        ],
        "domaineId": "m",
        "gestionnaire": null,
        "id": "arm",
        "typeId": "ar",
      },
    ],
    "typeId": "aut",
  },
  "administrationId": "aut-97300-01",
  "dateCreation": "2022-05-12",
  "email": "utilisateurEmail",
  "entreprises": Array [],
  "entreprisesCreation": true,
  "id": "utilisateurId",
  "modification": true,
  "motDePasse": "utilisateurMotdepasse",
  "newsletter": null,
  "nom": "utilisateurNom",
  "permissionModification": true,
  "preferences": null,
  "prenom": null,
  "refreshToken": null,
  "role": "editeur",
  "suppression": true,
  "telephoneFixe": null,
  "telephoneMobile": null,
  "utilisateursCreation": true,
}
`;

exports[`utilisateursQueryModify Vérifie l'écriture de la requête sur un utilisateur 2`] = `
Object {
  "administration": Object {
    "activitesTypes": Array [],
    "departementId": null,
    "id": "aut-97300-01",
    "regionId": null,
    "titresTypes": Array [
      Object {
        "archive": null,
        "associee": true,
        "contenuIds": Array [
          Object {
            "elementId": "mecanise",
            "sectionId": "arm",
          },
          Object {
            "elementId": "franchissements",
            "sectionId": "arm",
          },
        ],
        "domaineId": "m",
        "gestionnaire": null,
        "id": "arm",
        "typeId": "ar",
      },
    ],
    "typeId": "aut",
  },
  "administrationId": "aut-97300-01",
  "dateCreation": "2022-05-12",
  "email": "utilisateurEmail",
  "entreprises": Array [],
  "entreprisesCreation": true,
  "id": "utilisateurId",
  "modification": true,
  "motDePasse": "utilisateurMotdepasse",
  "newsletter": null,
  "nom": "utilisateurNom",
  "permissionModification": true,
  "preferences": null,
  "prenom": null,
  "refreshToken": null,
  "role": "editeur",
  "suppression": true,
  "telephoneFixe": null,
  "telephoneMobile": null,
  "utilisateursCreation": true,
}
`;

exports[`utilisateursQueryModify Vérifie l'écriture de la requête sur un utilisateur 3`] = `
Object {
  "administration": Object {
    "activitesTypes": Array [],
    "departementId": null,
    "id": "aut-97300-01",
    "regionId": null,
    "titresTypes": Array [
      Object {
        "archive": null,
        "associee": true,
        "contenuIds": Array [
          Object {
            "elementId": "mecanise",
            "sectionId": "arm",
          },
          Object {
            "elementId": "franchissements",
            "sectionId": "arm",
          },
        ],
        "domaineId": "m",
        "gestionnaire": null,
        "id": "arm",
        "typeId": "ar",
      },
    ],
    "typeId": "aut",
  },
  "administrationId": "aut-97300-01",
  "dateCreation": "2022-05-12",
  "email": "utilisateurEmail",
  "entreprises": Array [],
  "entreprisesCreation": true,
  "id": "utilisateurId",
  "modification": false,
  "motDePasse": "utilisateurMotdepasse",
  "newsletter": null,
  "nom": "utilisateurNom",
  "permissionModification": false,
  "preferences": null,
  "prenom": null,
  "refreshToken": null,
  "role": "editeur",
  "suppression": false,
  "telephoneFixe": null,
  "telephoneMobile": null,
  "utilisateursCreation": false,
}
`;

exports[`utilisateursQueryModify Vérifie l'écriture de la requête sur un utilisateur 4`] = `
Object {
  "administration": Object {
    "activitesTypes": Array [],
    "departementId": null,
    "id": "aut-97300-01",
    "regionId": null,
    "titresTypes": Array [
      Object {
        "archive": null,
        "associee": true,
        "contenuIds": Array [
          Object {
            "elementId": "mecanise",
            "sectionId": "arm",
          },
          Object {
            "elementId": "franchissements",
            "sectionId": "arm",
          },
        ],
        "domaineId": "m",
        "gestionnaire": null,
        "id": "arm",
        "typeId": "ar",
      },
    ],
    "typeId": "aut",
  },
  "administrationId": "aut-97300-01",
  "dateCreation": "2022-05-12",
  "email": "utilisateurEmail",
  "entreprises": Array [],
  "entreprisesCreation": false,
  "id": "utilisateurId",
  "modification": false,
  "motDePasse": "utilisateurMotdepasse",
  "newsletter": null,
  "nom": "utilisateurNom",
  "permissionModification": false,
  "preferences": null,
  "prenom": null,
  "refreshToken": null,
  "role": "editeur",
  "suppression": false,
  "telephoneFixe": null,
  "telephoneMobile": null,
  "utilisateursCreation": false,
}
`;

exports[`utilisateursQueryModify Vérifie l'écriture de la requête sur un utilisateur 5`] = `
Object {
  "administration": Object {
    "activitesTypes": Array [],
    "departementId": null,
    "id": "aut-97300-01",
    "regionId": null,
    "titresTypes": Array [
      Object {
        "archive": null,
        "associee": true,
        "contenuIds": Array [
          Object {
            "elementId": "mecanise",
            "sectionId": "arm",
          },
          Object {
            "elementId": "franchissements",
            "sectionId": "arm",
          },
        ],
        "domaineId": "m",
        "gestionnaire": null,
        "id": "arm",
        "typeId": "ar",
      },
    ],
    "typeId": "aut",
  },
  "administrationId": "aut-97300-01",
  "dateCreation": "2022-05-12",
  "email": "utilisateurEmail",
  "entreprises": Array [],
  "entreprisesCreation": false,
  "id": "utilisateurId",
  "modification": false,
  "motDePasse": "utilisateurMotdepasse",
  "newsletter": null,
  "nom": "utilisateurNom",
  "permissionModification": false,
  "preferences": null,
  "prenom": null,
  "refreshToken": null,
  "role": "editeur",
  "suppression": false,
  "telephoneFixe": null,
  "telephoneMobile": null,
  "utilisateursCreation": false,
}
`;
